src/v/datalake/coordinator/coordinator.cc

```diff
@@ -47,11 +54,147 @@ std::ostream& operator<<(std::ostream& o, coordinator::errc e) {
     }
 }
 
+void coordinator::start() {
+    ssx::spawn_with_gate(gate_, [this] { return run_until_abort(); });
+}
+
 ss::future<> coordinator::stop_and_wait() {
     as_.request_abort();
+    leader_cond_.broken();
     return gate_.close();
 }
 
+ss::future<> coordinator::run_until_abort() {
+    while (!as_.abort_requested()) {
+        auto& raft = *stm_->raft();
+        if (!raft.is_leader()) {
+            bool shutdown = false;
+            auto leader = co_await ss::coroutine::as_future(
+              leader_cond_.wait());
+            if (leader.failed()) {
+                auto eptr = leader.get_exception();
+                if (ssx::is_shutdown_exception(eptr)) {
+                    shutdown = true;
+                }
+            }
+            if (shutdown || as_.abort_requested()) {
+                co_return;
+            }
+        }
+        auto sync_res = co_await stm_->sync(10s);
+        if (sync_res.has_error()) {
+            auto msg = "Error while syncing coordinator";
+            switch (sync_res.error()) {
+            case coordinator_stm::errc::shutting_down:
+            case coordinator_stm::errc::not_leader:
+                vlog(datalake_log.debug, "{}", msg);
+                break;
+            case coordinator_stm::errc::raft_error:
+            case coordinator_stm::errc::apply_error:
+                vlog(datalake_log.warn, "{}", msg);
+                break;
+            }
+            continue;
+        }
+        auto synced_term = sync_res.value();
+        auto term_fut = co_await ss::coroutine::as_future(
+          run_until_term_change(synced_term));
+        if (term_fut.failed()) {
+            vlog(
+              datalake_log.error,
+              "Coordinator hit exception while running in term {}: {}",
+              synced_term,
+              term_fut.get_exception());
+            continue;
+        }
+        auto term_res = term_fut.get();
+        if (term_res.has_error()) {
+            vlog(
+              datalake_log.debug,
+              "Coordinator error while running in term {}: {}",
+              synced_term,
+              term_res.error());
+            continue;
+        }
+    }
+    co_return;
+}
+
+ss::future<checked<std::nullopt_t, coordinator::errc>>
+coordinator::run_until_term_change(model::term_id term) {
+    auto& raft = *stm_->raft();
+    vassert(!term_as_.has_value(), "Multiple calls to run_until_term_change");
+    ss::abort_source term_as;
+    term_as_ = term_as;
```

Comment:
term_as_ seems to only be used in this one coroutine, so does it need to be a class member?